packages/monaco/uplift-journal.md

```diff
@@ -0,0 +1,125 @@
+
+### Observations re: colorization
+Language identification does fail on startup (all editors other than JSONC are incorrectly identified as `plaintext`), but events appear to be hooked up correctly so that editor languages are updated as plugins are processed, and colorization even fails for JSONC. It must not be a problem with mime-types, but something to do with colorization / tokenization itself.
+
+In the end, it turned out that we had just missed a change in the type of a particular field, and it wasn't nicely typed in the declaration file. I'm not sure how it wasn't throwing an error: apparently `Object.keys` always returns an empty array on a `Set`, so we were never trying to call methods that didn't exist; we just also weren't doing anything at all.
+
+### Setting definite version
+
+I pulled the 1.64.2 tag of VSCode today to pin my implementations to that. It is a couple of weeks behind what I'd pulled from `main`, and in n just that couple of weeks,a number of private API references were broken. We really shouldn't rely on private API.
+
+## Tweaking monaco-editor-core
+
+In the end, some additional tweaks were necessary to get everything working with `monaco-editor-core`. We compile to `commonjs` modules, and `monaco-editor-core` uses ESM. In the application, that's fine: we only use Monaco in the browser, and Webpack handles the interface between the two styles. In our tests, however, we're running in a Node environment without packing, so `require` calls that refer to Monaco code fail. To fix that, I made the Monaco build output `commonjs` instead of ESM by changing the line in `vscode/build/lib/standalone.js` that sets `module = 'es6'` to `module = 'commonjs'`. In addition, I had to add a `main` field in the `package.json` to help `require('monaco-editor-core')` calls find the right entry point. Finally, Monaco ships built versions of two dependencies, `marked` and `dompurify` as `.js` files in ESM style, and since those don't get built with the rest of the code, they didn't undergo the same change to `commonjs`. In those cases, I just manually rewrote the exports to `commonjs` style in the generated code. Those files are checked in at `src/vs/base/browser/dompurify/dompurify.js` and `src/vs/base/common/marked/marked.js`, so the changes can also be made in the 'source'.
```

Comment:
I got consuming es modules from commonjs to work by using dynamic imports: the problem is that typescript compiles those to "require", which is not allowed. I got around it by writing:

```
eval("import('somemodule')");
```

instead of 

```
import('somemodule');
```

Maybe typescript needs a `#VERBATIM` pragma to just pass through some javascript without compilation.

Reply:
Since dynamic `import` is asynchronous and `require` is synchronous, I'm not sure that working around it that way would work consistently, especially since CommonJS doesn't have top-level `await`?